### ISSUE TITLE
SLEIGH: offset attribute when naming registers

### ISSUE THREAD

**Issue by jroimartin** [NationalSecurityAgency/ghidra#831]
Let's say that we define the following address space:

```
define space mem type=ram_space size=2 wordsize=2;
```

On the other hand, we also define the following register within that space:

```
define mem offset=0x2000 size=2 [ MYREG ];
```

Now, my question is: when naming the register, does the offset attribute take into account the wordsize of the space? I mean, what will be the address of MYREG? mem:0x2000 or mem:0x1000?

In the code browser when I double click (or hover) the register (which is attached to a field) it points me to the address 0x1000. So, it seems that the offset does not take into account the wordsize of the space and uses a wordsize of 1. However, I cannot use an offset bigger than the size of the space (2 bytes in the example), so how would I define a register at the very end of the space (e.g. mem:0xfffe)?

**Comment by mumbel** [comment]
>Each name in the list is assigned to a varnode in turn starting at the indicated offset within the space. Each varnode occupies the indicated number of bytes in size.

MYREG is at 0x2000 and takes up 2 bytes.  Are you sure your project/ghidra/sleigh all sync'd up?

**Comment by jroimartin** [comment]
Yes, I think everything is in sync. Actually, if I multiply the offset by 2 (offset=0x4000, in the previous example) I get the expected result. Meaning that double-clicking the register goes to mem:0x2000.

**Comment by jroimartin** [comment]
To make things a bit more clear, I wrote the following minimal sleigh specification:

```
define endian=little;
define alignment=1;

define space code type=ram_space size=2 default;
define space mem type=ram_space size=2 wordsize=2;

define space register type=register_space size=2;

define register offset=0x0 size=2 [
        PC SP r2
];

define mem offset=0x10 size=2 [ REG0 REG1 ];

define token instr8(8)
        op  = (1,7)
        reg = (0,0)
;

attach variables [ reg ] [ REG0 REG1 ];

:FOO reg is op=0 & reg unimpl
```

Disassembling a file containing `\x00` produces the following result:

![image](https://user-images.githubusercontent.com/1223476/61861680-b5502280-aecc-11e9-89f7-fd674cbeb173.png)

As you can see, in the sleigh spec the offset of `REG0` is `0x10`. On the other hand, the code browser shows `REG0` in address `mem:0008` which corresponds to `offset / wordsize`.

Now, let's say I want REG0 to be in address `mem:9000`. That means that I should change my sleigh spec in the following way:

```
# offset = 0x9000 * wordsize = 0x9000 * 2 = 0x12000
define mem offset=0x12000 size=2 [ REG0 REG1 ];
```

However, if I do so, the compilation fails with the following error:

```
ERROR skel.slaspec:13: offset 0x12000 is too large for space mem (SleighCompile)  
```

What's the right thing to do in this case?

**Comment by mumbel** [comment]
Good example showing the issue, that does clear it up.  So offset is in bytes, but when reading/writing sleigh it definitely makes sense, from one POV, to be in wordsize.  What processor are you working on?

(@ryanmkurtz  should this be relabeled as bug?)

**Comment by ghidra1** [comment]
This appears to be a bug in the Java version of the Sleigh compiler.  This will be corrected for 9.1.  You can work around this bug by compiling with the native Sleigh compiler located in Decompiler module (Ghidra/Features/Decompiler/os/.../sleigh).

**Comment by ryanmkurtz** [comment]
Fixed [here](https://github.com/NationalSecurityAgency/ghidra/commit/6d4908f0cdb32b0cbd3954c4a416e721a168e54e).
